### PR TITLE
ssh/tailssh: also handle recording upload failure during writes

### PR DESCRIPTION
**Pull request by maisem**
Previously we would error out when the recording server disappeared after the in memory buffer filled up for the io.Copy. This makes it so that we handle failing open correctly in that path.

Updates tailscale/corp#9967

### CODE REVIEW

File: ssh/tailssh/tailssh.go [hunk 1]
```diff
@@ -1519,8 +1519,9 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
 
 	now := time.Now()
 	rec := &recording{
-		ss:    ss,
-		start: now,
+		ss:       ss,
+		start:    now,
+		failOpen: onFailure == nil || onFailure.TerminateSessionWithMessage == "",
```
Comment (twitchyliquid64) [inline ssh/tailssh/tailssh.go:1524]: Why this second conditional? Can we make a nil callback be how fail open is communicated?

Comment (maisem) [reply ssh/tailssh/tailssh.go:1524]: Nil callback is useful for when the recording fails to start. This is for when it starts and then fails.

File: ssh/tailssh/tailssh.go [hunk 2]
```diff
@@ -1642,7 +1647,7 @@ func (r *recording) writer(dir string, w io.Writer) io.Writer {
 		// passwords.
 		return w
 	}
-	return &loggingWriter{r, dir, w}
+	return &loggingWriter{r, dir, w, false}
```
Comment (catzkorn) [inline ssh/tailssh/tailssh.go:1650]: Can you name these 😅

**Review by catzkorn — APPROVED** [review]

File: ssh/tailssh/tailssh.go (hunk 2 above)
Comment (maisem) [reply ssh/tailssh/tailssh.go]: Done

File: ssh/tailssh/tailssh.go [hunk 3]
```diff
@@ -1651,20 +1656,29 @@ type loggingWriter struct {
 	r   *recording
 	dir string    // "i" or "o" (input or output)
 	w   io.Writer // underlying Writer, after writing to r.out
+
+	// recordingFailed is whether we should skip recording.
+	// It is set to true if we fail to write to r.out and r.failOpen is set.
+	recordingFailed bool
```
Comment (creachadair) [inline ssh/tailssh/tailssh.go:1662]: Minor thing: Maybe `recordingFailedOpen` just to reinforce that this means not only the failure but also that we checked openness?

Comment (maisem) [reply ssh/tailssh/tailssh.go]: good point. done

**Pull request merged**